src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/DictionarySlim.cs

```diff
@@ -0,0 +1,382 @@
+
+        public TValue GetValueOrDefault(TKey key)
+        {
+            Entry[] entries = _entries;
+            int entryIndex = GetEntryIndex((int)(GetHashCode(key) % (uint)entries.Length));
+
+            while (entryIndex != -1)
+            {
+                if (entries[entryIndex].key.Equals(key))
+                {
+                    return entries[entryIndex].value;
+                }
+                entryIndex = entries[entryIndex].next;
+            }
+
+            return default;
+        }
+
+        public bool Remove(TKey key)
+        {
+            Entry[] entries = _entries;
+            int bucketIndex = (int)(GetHashCode(key) % (uint)entries.Length);
+            int entryIndex = GetEntryIndex(bucketIndex);
+            
+            int lastIndex = -1;
+            while (entryIndex != -1)
+            {
+                if (entries[entryIndex].key.Equals(key))
+                {
+                    if (lastIndex != -1)
+                    {   // Fixup preceding element in chain to point to next (if any)
+                        entries[lastIndex].next = entries[entryIndex].next;
+                    }
+                    else
+                    {   // Fixup bucket to new head (if any)
+                        _buckets[bucketIndex] = entries[entryIndex].next + 1;
+                    }
+
+                    entries[entryIndex] = default; // could use RuntimeHelpers.IsReferenceOrContainsReferences
+
+                    entries[entryIndex].next = _freeList; // New head of free list
+                    _freeList = entryIndex;
+
+                    Count--;
+                    return true;
+                }
+                lastIndex = entryIndex;
+                entryIndex = entries[entryIndex].next;
+            }
+
+            return false;
+        }
+
+        public ref TValue this[TKey key]
+        {
+            get
+            {
+                Entry[] entries = _entries;
+                int bucketIndex = (int)(GetHashCode(key) % (uint)entries.Length);
+                int entryIndex = GetEntryIndex(bucketIndex);
+
+                for(int i = entryIndex; i != -1; i = entries[i].next)
+                {
+                    if (entries[i].key.Equals(key))
+                    {
+                        return ref entries[i].value;
+                    }
+                }
+
+                if (_freeList != -1)
+                {
+                    entryIndex = _freeList;
+                    _freeList = entries[_freeList].next;
+                }
+                else
+                {
+                    if (Count == entries.Length)
+                    {
+                        entries = Resize();
+                        bucketIndex = (int)(GetHashCode(key) % (uint)entries.Length);
+                        // entry indexes were not changed by Resize
+                    }
+                    entryIndex = Count;
+                }
+
+                entries[entryIndex].key = key;
```

Comment:
Consider using a ref variable for `entries[entryIndex]`.

Reply:
Interesting, not what I would have expected.

Reply:
JIT's CSE is usually capable of handling this kind of repeated accesses so it's not surprising that you don't see an improvement. That said, I don't remember exactly why it's slower. For reference arrays you have to be careful with taking the address of an element as you may walk straight into the array covariance swamp. For struct arrays, hmm, it's possible that taking the address interferes with address mode formation and leaves you with a rather expensive 3 operand LEA instruction.

Reply:
Cool. Will keep trying anyway :-).